### PR TITLE
properly handle Join requests, fix all tests.

### PR DESCRIPTION
**Pull request by krasi-georgiev**
The comment here explains the proper way that join requests work with the rak811 module

https://github.com/calvernaz/rak811/compare/master...krasi-georgiev:join-fix?expand=1#diff-8970630c2d7366c16aa08aaae509aa87R190-R195

Realized that after a lot of troubleshooting :smiley_cat: 

Signed-off-by: Krasi Georgiev <8903888+krasi-georgiev@users.noreply.github.com>

### CODE REVIEW

**Comment by calvernaz** [comment]
There is are conflicts with some recent changes I did, so please update your branch.

**Review by calvernaz — COMMENTED** [review]
The check for the "OK" is not required, the previous version had an issue handling some responses, this commit improves on that, but the truth is we don't need to check for the "OK" string.

We read until there is nothing to read and the caller is responsible for parsing the response.

File: rak811.go [hunk 1]
```diff
@@ -68,7 +68,7 @@ func (l *Lora) tx(cmd string) (string, error) {
 		}
 		log.Fatalf("reading response err:%v", err)
 	}
-	return "", fmt.Errorf("unexpected response:%v", resp)
+	return "", fmt.Errorf("unexpected tx response:%v cmd:%v", resp, cmd)
```
Comment (calvernaz) [inline rak811.go]: I also did some changes on the error wrapping, that is aligned with the recent changes for Golang 1.13

File: rak811.go [hunk 2]
```diff
@@ -57,7 +57,7 @@ func (l *Lora) tx(cmd string) (string, error) {
 	for scanner.Scan() {
 		t := scanner.Text()
 		resp += t
-		if t == "OK" {
+		if strings.HasPrefix(t, "OK") {
```
Comment (calvernaz) [inline rak811.go]: The check for the "OK" is not required, the previous version had an issue handling some responses, this commit improves on that, but the truth is we don't need to check for the "OK" string.

We read until there is nothing to read and the caller is responsible for parsing the response

File: rak811_test.go [hunk 3]
```diff
@@ -407,27 +369,46 @@ func TestLora_GetRadioStatus(t *testing.T) {
...
+	if err != nil {
+		return 0, err
+	}
+	n = copy(p, line)
+	f.last = line
 	return n, nil
 }
 
-func (FakeSerialPort) Write(p []byte) (n int, err error) {
+func (*FakeSerialPort) Write(p []byte) (n int, err error) {
```
Comment (calvernaz) [inline rak811_test.go]: There is no need for point receiver if it's not modifiying

File: rak811_test.go [hunk 4]
```diff
@@ -407,27 +369,46 @@ func TestLora_GetRadioStatus(t *testing.T) {
...
+
 type FakeSerialPort struct {
-	buf []byte
+	reader *bufio.Reader
+	last   []byte
 }
 
-func (f FakeSerialPort) Read(p []byte) (n int, err error) {
-	n = copy(p, f.buf)
+func (f *FakeSerialPort) Read(p []byte) (n int, err error) {
```
Comment (calvernaz) [inline rak811_test.go:390]: Not sure why is this required, is this related to the Join change?

File: rak811.go [hunk 5]
```diff
@@ -162,40 +163,64 @@ func (l *Lora) SetConfig(config string) (string, error) {
...
-	resp, err := bufio.NewReader(l.port).ReadString('\n')
-	if err != nil {
-		if err == io.EOF {
-			return "", ErrTimeout
+	var (
+		resp  string
+		err   error
+		start = time.Now()
+	)
+	for {
```
Comment (calvernaz) [inline rak811.go:195]: the reader should be created once outside the loop

File: rak811_test.go (hunk 3 above)
Comment (krasi-georgiev) [reply rak811_test.go]: True, but common golang standard is that if one receiver is a pointer all receivers should be pointers.
In this case the Read method needs to be a pointer receiver.

File: rak811_test.go (hunk 4 above)
Comment (krasi-georgiev) [reply rak811_test.go:390]: Sorry, I didn't understand the question.

File: rak811.go (hunk 2 above)
Comment (krasi-georgiev) [reply rak811.go]: good idea, refactored.

File: rak811.go (hunk 5 above)
Comment (krasi-georgiev) [reply rak811.go:195]: updated

File: rak811.go (hunk 1 above)
Comment (krasi-georgiev) [reply rak811.go]: Aaah , nice, just had a quick look.

What changes do you suggest here?

File: rak811_test.go (hunk 3 above)
Comment (calvernaz) [reply rak811_test.go]: This should be used as the guideline and is not that important to me in this situation, but for future reference, we should consider this list instead of just because all some receivers use pointers. 

https://github.com/golang/go/wiki/CodeReviewComments#receiver-type

File: rak811.go (hunk 1 above)
Comment (calvernaz) [reply rak811.go]: Until 1.13 I believe we can't use the  `: %w` format. 

You can read more [here](https://github.com/golang/go/wiki/ErrorValueFAQ#i-am-already-using-fmterrorf-with-v-or-s-to-provide-context-for-an-error-when-should-i-switch-to-w)

A detail but an important one, leave the error as the last argument and the format with a space after the colon `: %v`.
